daemon/c_seccomp.c

```diff
@@ -0,0 +1,534 @@
+
+		/* [2] Load system call number from 'seccomp_data' buffer into
+                accumulator */
+		BPF_STMT(BPF_LD | BPF_W | BPF_ABS, (offsetof(struct seccomp_data, nr))),
+
+		/* [3] Check ABI - only needed for x86-64 in blacklist use
+                cases.Use BPF_JGT instead of checking against the bit
+                mask to avoid having to reload the syscall number. */
+		BPF_JUMP(BPF_JMP | BPF_JGT | BPF_K, upper_nr_limit, 4, 0),
+
+		/* [4] Jump forward 1 instruction if system call number
+                does not match 'syscall_nr' */
+		BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, SYS_clock_settime, NCALLS - 1, 0),
+		BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, SYS_clock_adjtime, 0, 1),
+
+		/* [5] Matching architecture and system call: don't execute
+            the system call, and return 'f_errno' in 'errno' */
+		//BPF_STMT(BPF_RET | BPF_K,
+		//      SECCOMP_RET_ERRNO | (f_errno & SECCOMP_RET_DATA)),
+		BPF_STMT(BPF_RET + BPF_K, SECCOMP_RET_USER_NOTIF),
+
+		/* [6] Destination of system call number mismatch: allow other
+                system calls */
+		BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
+
+		/* [7] Destination of architecture mismatch: kill task */
+		BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_KILL),
+	};
+
+	struct sock_fprog prog = { .len = (unsigned short)(sizeof(filter) / sizeof(filter[0])),
+				   .filter = filter };
+
+	int ret;
+	if (-1 ==
+	    (ret = seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_NEW_LISTENER, &prog))) {
+		ERROR("SECCOMP_SET_MODE_FILTER: ");
+		ERROR("Return value was %d", ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+static int
+c_seccomp_start_pre_exec_child_early(void *seccompp)
+{
+	ASSERT(seccompp);
+	c_seccomp_t *seccomp = seccompp;
+
+	int notify_fd = -1;
+
+	if (!seccomp_is_pidfd_supported()) {
+		INFO("Kernel does not support pidfd, skipping seccomp_notify-based syscall emulation");
+		return 0;
+	}
+
+	if (-1 == (notify_fd = install_filter(AUDIT_ARCH_X86_64, EPERM))) {
+		ERROR("Failed to install filter for syscall %d\n", SYS_creat);
+		return -1;
+	}
+
+	DEBUG("Installed seccomp filter, notify fd is %d\n", notify_fd);
+
+	DEBUG("sending seccomp notify fd number to parent");
+
+	if (fd_write(compartment_get_sync_sock_child(seccomp->compartment), (char *)&notify_fd,
+		     sizeof(notify_fd)) < 0) {
+		WARN_ERRNO("Could not send notify fd number over sync socket");
+		return -1;
+	}
+
+	compartment_set_notify_fd(seccomp->compartment, notify_fd);
+
+	return 0;
+}
+
+void *
+c_seccomp_fetch_vm_new(int pid, void *rbuf, uint64_t size)
+{
+	IF_NULL_RETVAL(rbuf, NULL);
+	IF_TRUE_RETVAL(pid < 0, NULL);
+
+	uint64_t bytes_read = 0;
+	void *lbuf = mem_alloc0(size);
+	struct iovec local_iov[1];
+	struct iovec remote_iov[1];
+
+	local_iov[0].iov_base = lbuf;
+	local_iov[0].iov_len = size;
+
+	remote_iov[0].iov_base = rbuf;
+	remote_iov[0].iov_len = size;
+
+	if (size !=
+	    (bytes_read = syscall(SYS_process_vm_readv, pid, local_iov, 1, remote_iov, 1, 0))) {
```

Comment:
put this above if and then just
if (size != bytes_read) {

Reply:
done